metadata-ingestion/src/datahub/ingestion/source/vertexai.py

```diff
@@ -0,0 +1,497 @@
+)
+@capability(SourceCapability.TAGS, "Extract tags for VertexAI Registered Model Stages")
+class VertexAISource(Source):
+    platform = "vertexai"
+
+    def __init__(self, ctx: PipelineContext, config: VertexAIConfig):
+        super().__init__(ctx)
+        self.config = config
+        self.report = SourceReport()
+        aiplatform.init(project=config.project_id, location=config.region)
+        self.client = aiplatform
+        self.endpoints = None
+
+    def get_report(self) -> SourceReport:
+        return self.report
+
+    def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
+        """
+        Main Function to fetch and yields work units for various VertexAI resources.
+        - Models and Model Versions from the Model Registry
+        - Training Jobs
+        """
+        # Fetch Models, Model Versions a from Model Registry
+        yield from self._get_ml_model_workunits()
+        # Fetch Training Jobs
+        yield from self._get_training_job_workunit()
+        # TODO Fetch Experiments and Experiment Runs
+
+    def _create_workunit(self, urn: str, aspect: _Aspect) -> MetadataWorkUnit:
+        """
+        Utility to create an MCP workunit.
+        """
+        return MetadataChangeProposalWrapper(
+            entityUrn=urn,
+            aspect=aspect,
+        ).as_workunit()
+
+    def _validate_training_job(self, model: Model) -> bool:
+        """
+        Validate Model Has Valid Training Job
+        """
+        job = model.training_job
+        if not job:
+            return False
+
+        try:
+            # when model has ref to training job, but field is not accessible, it is not valid
+            name = job.name
+            logger.debug((f"can fetch training job name: {name} for model: (name:{model.display_name} id:{model.name})"))
+            return True
+        except RuntimeError:
+            logger.debug(f"cannot fetch training job name, not valid for model (name:{model.display_name} id:{model.name})")
+
+        return False
+
+    def _get_ml_model_workunits(self) -> Iterable[MetadataWorkUnit]:
+        """
+        Fetch List of Models in Model Registry and generate a corresponding work unit.
+        """
+        registered_models = self.client.Model.list()
+        for model in registered_models:
+            # create work unit for Model Group (= Model in VertexAI)
+            yield self._get_ml_group_workunit(model)
+            model_versions = model.versioning_registry.list_versions()
+            for model_version in model_versions:
+
+                # create work unit for Training Job (if Model has reference to Training Job)
+                if self._validate_training_job(model):
+                    logger.info(
+                        f"Generating TrainingJob work unit for model: {model_version.model_display_name}")
+                    yield from self._get_data_process_properties_workunit(model.training_job)
```

Comment:
A: model sometimes has a training job

B: training jobs usually have a model

if A, then B?
if B, then A?

when does model -> training job exist, but training job -> model not?



Harshal's suggestion
1. iterate +generate all training jobs
2. keep a mapping from model -> training job while iterating
3. iterate + generate models; if we find a model with an associated training job (either via A, or via B's mapping)

Reply:
if the mapping turns out to be massive, use a `FileBackedDict`